src/core/Akka.Persistence/Fsm/PersistentFSM.cs

```diff
@@ -131,7 +131,8 @@ void ApplyStateOnLastHandler()
                     handlersExecutedCounter++;
                     if (handlersExecutedCounter == eventsToPersist.Count)
                     {
-                        base.ApplyState(nextState.Using(nextData));
+                        nextState = nextState.Using(nextData);
```

Comment:
Isn't this a change in behavior, since nextState is actually set downstream in the MakeTransition method? Looking a the JVM, what we're missing is:

```csharp
base.ApplyState(nextState.Copy(stateData: nextData));
```

Reply:
I must be missing something. Here's what's happening now in ApplyStateOnLastHandler which is called after the persistence is completed.

https://github.com/akkadotnet/akka.net/blob/d9f6c573afe1ba59c3d6733dda4bc1f7dd089efc/src/core/Akka.Persistence/Fsm/PersistentFSM.cs#L129-L143


With my comments:
```
 void ApplyStateOnLastHandler()
{
    handlersExecutedCounter++;
    if (handlersExecutedCounter == eventsToPersist.Count)
    {
        // Using creates a COPY of nextState with stateData set to nextData
        // nextState still refers to the OLD version WITHOUT the NEW nextData being set
        base.ApplyState(nextState.Using(nextData));

        // uses OLD nextState, but that doesn't matter, as Timeout hasn't changed        
        CurrentStateTimeout = nextState.Timeout;

        // HERE's the issue and the actual bug
        // The "AfterTransitionDo" callback (ie, AndThen) - if present - is
        // called with the StateData Property of the OLD, UNCHANGED state.
        // That is, this StateData is always lagging one version behind.        
        nextState.AfterTransitionDo?.Invoke(nextState.StateData);

        // the rest is not relevant
        if (doSnapshot)
        {
            Log.Info($"Saving snapshot, sequence number [{SnapshotSequenceNr}]");
            SaveStateSnapshot();
        }
    }
}
```

This issue was not discovered using the current set of unit tests, but it is very easily to show (eg. in the shopping cart example) that the current version is *broken*.

Please try it for yourself. Here's a snippet from the specs:
```
When(LookingAround.Instance, (evt, state) =>
{
    // NB: When in this state, the cart is empty
    if (evt.FsmEvent is AddItem addItem)
    {
        // visitor adds an item to the EMPTY cart
        return GoTo(Shopping.Instance)
            .Applying(new ItemAdded(addItem.Item))
            .ForMax(TimeSpan.FromSeconds(1))
            .AndThen(cart => /* what to you expect cart to be? Empty or non-empty? */);
    }
    [...]
});
```

The issue is at the "/* what to you expect cart to be? Empty or non-empty? */" part. You'd expect the cart to be non-empty, but it isn't. It's still the OLD, EMPTY version.

Reply:
So I checked, and line [#136](https://github.com/akkadotnet/akka.net/blob/ad59e30ac3058a0c22301214b526a00011486c8c/src/core/Akka.Persistence/Fsm/PersistentFSM.cs#L137) needs to be fixed too:

```csharp
nextState.AfterTransitionDo?.Invoke(StateData);
```
With these two changes, it works as expected.

Reply:
It should actually work without. To me, it's either-or. My PR is one solution. The other one is `nextState.AfterTransitionDo?.Invoke(nextData);` which fixes the issue, too, but for my personal taste is the less clean approach. Ymmv.

Reply:
Ok, the jvm version uses the second approach. I'll stick to that and will update the PR asap.